File: android/app/src/main/java/info/bagen/rust/plaoc/microService/ipc/IpcBodyReceiver.kt
package info.bagen.rust.plaoc.microService.ipc

import info.bagen.rust.plaoc.microService.helper.SIGNAL_CTOR
import info.bagen.rust.plaoc.microService.helper.asBase64
import info.bagen.rust.plaoc.microService.helper.asUtf8
import info.bagen.rust.plaoc.microService.helper.debugger
import java.io.InputStream


/**
 * metaBody 可能会被多次转发，
 * 但只有第一次得到这个 metaBody 的 ipc 才是它真正意义上的 Receiver
 */
class IpcBodyReceiver(
    override val metaBody: MetaBody,
    ipc: Ipc,
) : IpcBody() {

    class IPC {
        companion object {

        }
    }

    /// 因为是 abstract，所以得用 lazy 来延迟得到这些属性
    override val bodyHub by lazy {
        BodyHub().also {
            val data = if (metaBody.type.isStream) {
                val ipc = CACHE.metaId_receiverIpc_Map[metaBody.metaId]
                    ?: throw Exception("no found ipc by metaId:${metaBody.metaId}")
                metaToStream(metaBody, ipc)
            } else when (metaBody.type.encoding) {
                /// 文本模式，直接返回即可，因为 RequestInit/Response 支持支持传入 utf8 字符串
                IPC_DATA_ENCODING.UTF8 -> metaBody.data as String
                IPC_DATA_ENCODING.BINARY -> metaBody.data as ByteArray
                IPC_DATA_ENCODING.BASE64 -> (metaBody.data as String).asBase64()
                else -> throw Exception("invalid metaBody type: ${metaBody.type}")
            }
            it.data = data
            when (data) {
                is String -> it.text = data;
                is ByteArray -> it.u8a = data
                is InputStream -> it.stream = data
            }
        }
    }

    init {
        /// 将第一次得到这个metaBody的 ipc 保存起来，这个ipc将用于接收
        if (metaBody.type.isStream) {
            CACHE.metaId_receiverIpc_Map.getOrPut(metaBody.metaId) {
                ipc.onClose {
                    CACHE.metaId_receiverIpc_Map.remove(metaBody.metaId)
                }
                metaBody.receiverUid = ipc.uid
                ipc
            }
        }
    }

    companion object {

        fun from(metaBody: MetaBody, ipc: Ipc): IpcBody {
            return CACHE.metaId_ipcBodySender_Map[metaBody.metaId] ?: IpcBodyReceiver(metaBody, ipc)
        }


        /**
         * @return {String | ByteArray | InputStream}
         */
        fun metaToStream(metaBody: MetaBody, ipc: Ipc): InputStream {
            /// metaToStream
            val stream_id = metaBody.streamId!!;
            val stream = ReadableStream(cid = "receiver-${stream_id}", onStart = { controller ->

                /// 如果有初始帧，直接存起来
                when (metaBody.type.encoding) {
                    IPC_DATA_ENCODING.UTF8 -> (metaBody.data as String).asUtf8()
                    IPC_DATA_ENCODING.BINARY -> metaBody.data as ByteArray
                    IPC_DATA_ENCODING.BASE64 -> (metaBody.data as String).asBase64()
                    else -> null
                }?.let { firstData -> controller.enqueue(firstData) }

                ipc.onMessage { (message) ->
                    if (message is IpcStreamData && message.stream_id == stream_id) {
                        debugIpcBody(
                            "receiver/StreamData/$ipc/${controller.stream}", message
                        )
                        if (stream_id == "rs-0") {
                            debugger()
                        }
                        controller.enqueue(message.binary)
                    } else if (message is IpcStreamEnd && message.stream_id == stream_id) {
                        debugIpcBody(
                            "receiver/StreamEnd/$ipc/${controller.stream}", message
                        )
                        controller.close()
                        return@onMessage SIGNAL_CTOR.OFF
                    } else {
                    }
                }
            }, onPull = { (desiredSize, controller) ->
                debugIpcBody(
                    "receiver/postPullMessage/$ipc/${controller.stream}", stream_id
                )
                ipc.postMessage(IpcStreamPull(stream_id, 1))
            });

            debugIpcBody("receiver/$ipc/$stream", "start by stream-id:${stream_id}")

            return stream

        }
    }
}

File: ios/BFExplorer/Plaoc/bfs/ipc/IpcBodyReceiver.swift
//
//  IpcBodyReceiver.swift
//  BFExplorer
//
//  Created by ui08 on 2023/2/27.
//

import Foundation
import Vapor

class IpcBodyReceiver: IpcBody {
    private let ipc: Ipc
    
    init(metaBody: MetaBody, ipc: Ipc) {
        self.ipc = ipc
        super.init()
        self.metaBody = metaBody
        
        if self.metaBody.type.isStream {
            let cacheIpc = CACHE.metaId_receiverIpc_Map[self.metaBody.metaId]
            
            if cacheIpc == nil {
                _ = ipc.onClose {
                    CACHE.metaId_receiverIpc_Map.removeValue(forKey: self.metaBody.metaId)
                    return .OFF
                }
                self.metaBody.receiverUid = ipc.uid
                
                CACHE.metaId_receiverIpc_Map[self.metaBody.metaId] = ipc
            }
        }
    }
    
    private lazy var _bodyHub: IpcBody.BodyHub = {
        var bodyHub = BodyHub()

        if self.metaBody.type.isStream {
            bodyHub.stream = IpcBodyReceiver.metaToStream(metaBody: self.metaBody, ipc: self.ipc)
            bodyHub.data = bodyHub.stream
        } else {
            switch self.metaBody.type.encoding {
            case .utf8:
                bodyHub.text = self.metaBody.data.string!
                bodyHub.data = bodyHub.text
            case .binary:
                bodyHub.u8a = self.metaBody.data.data!
                bodyHub.data = bodyHub.u8a
            case .base64:
                bodyHub.u8a = self.metaBody.data.string!.fromBase64()!
                bodyHub.data = bodyHub.u8a
            default:
                fatalError("invalid metaBody type: \(self.metaBody.type)")
            }
        }
        
//        switch data {
//        case is String:
//            bodyHub.text = data as! String
//        case is Data:
//            bodyHub.u8a = data as! Data
//        case is InputStream:
//            bodyHub.stream = data as! InputStream
//        default:
//            fatalError("invalid data type: \(data)")
//        }
        return bodyHub
    }()
    override var bodyHub: IpcBody.BodyHub {
        get {
            _bodyHub
        }
        set {
            _bodyHub = newValue
        }
    }
    
    static func from(metaBody: MetaBody, ipc: Ipc) -> IpcBody {
        return CACHE.metaId_ipcBodySender_Map[metaBody.metaId] ?? IpcBodyReceiver(metaBody: metaBody, ipc: ipc)
    }
    
    /**
     * @return {String | ByteArray | InputStream}
     */
    static func metaToStream(metaBody: MetaBody, ipc: Ipc) -> InputStream {
        var metaBody = metaBody
        let stream_id = metaBody.streamId!
        let stream = ReadableStream(onStart: { controller in
            var data: Data?
            switch metaBody.type.encoding {
            case .utf8:
                data = metaBody.data.string!.fromUtf8()
            case .binary:
                data = metaBody.data.data!
            case .base64:
                data = metaBody.data.string!.fromBase64()
            default:
                data = nil
            }
            
            /// 如果有初始帧，直接存起来
            if data != nil {
                controller.enqueue(data!)
            }
            
            _ = ipc.onMessage { (message, _) in
                if var message = message as? IpcStreamData, message.stream_id == stream_id {
                    if stream_id == "rs-0" {
                        print("ReadableStream stream_id: \(stream_id)")
                    }
                    controller.enqueue(message.binary)
                } else if let message = message as? IpcStreamEnd, message.stream_id == stream_id {
                    print("receiver/StreamEnd/\(ipc)/\(controller.stream)", message)
                    controller.close()
                    
                    return SIGNAL_CTOR.OFF
                }
                
                return nil
            }
        }, onPull: { (desiredSize, controller) in
            print("receiver/postPullMessage/\(ipc)/")
            await ipc.postMessage(message: IpcStreamPull(stream_id: stream_id, desiredSize: 1))
        })
        
        print("receiver/\(ipc)/\(stream)", "start by stream-id:\(stream_id)")
        
        return stream
    }
}
